core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,241 @@
+from trezor.crypto import hashlib
+from trezor.crypto.curve import ed25519
+from trezor.messages import CardanoAddressType
+
+from apps.common import cbor
+
+from .address import derive_address_bytes, derive_human_readable_address
+from .helpers import INVALID_AUXILIARY_DATA, bech32
+from .helpers.bech32 import HRP_JORMUN_PUBLIC_KEY
+from .helpers.paths import SCHEMA_STAKING_ANY_ACCOUNT
+from .helpers.utils import derive_public_key
+from .layout import confirm_catalyst_registration, show_auxiliary_data_hash
+
+if False:
+    from typing import Union
+    from trezor import wire
+
+    from trezor.messages.CardanoCatalystRegistrationParametersType import (
+        CardanoCatalystRegistrationParametersType,
+    )
+    from trezor.messages.CardanoTxAuxiliaryDataType import CardanoTxAuxiliaryDataType
+
+    CatalystRegistrationPayload = dict[int, Union[bytes, int]]
+    CatalystRegistrationSignature = dict[int, bytes]
+    CatalystRegistration = dict[
+        int, Union[CatalystRegistrationPayload, CatalystRegistrationSignature]
+    ]
+
+    from . import seed
+
+AUXILIARY_DATA_HASH_SIZE = 32
+CATALYST_VOTING_PUBLIC_KEY_LENGTH = 32
+CATALYST_REGISTRATION_HASH_SIZE = 32
+
+METADATA_KEY_CATALYST_REGISTRATION = 61284
+METADATA_KEY_CATALYST_REGISTRATION_SIGNATURE = 61285
+
+
+def validate_auxiliary_data(
```

Comment:
this should also check that not more than one of the fields is set

proposal for a simple check:
```
data_provided = 0
if blob:
    data_provided += 1
    ...
elif params:
    data_provided += 1
    ...

...
if data_provided != 1:
    raise
```

Reply:
Sure. I wanted to write a device test for this, but the parsing in trezorlib has the same if/elif structure, so it only includes one of the fields. Should we maybe also update `python/src/trezorlib/cardano.py` to include both fields if they are passed in? Or should trezorlib also validate that only one of the fields is present?

Or maybe the client libraries (trezorlib/connect) should be the ones validate this? Because this doesn't really result in an invalid transaction, since the serialization does a if/elif for the fields.

Reply:
We don't want to accept the bad data on the device side, because then it's unclear which will be used if you send both.

You can update trezorlib to do either; raising an exception is nicer for the end-user, including both fields would allow us to test the device behavior. I don't consider it trezorlib's job to validate data this deep so I don't have a strong opinion about what happens.

Regarding that, ISTM the main reason most of the code in `trezorlib/cardano.py` exists at all is to be able to call `parse_path` on paths that are deep inside the nested structure. Instead, we could get rid of it and use `protobuf.dict_to_proto`, if we added smart handling when expected field is `repeated uint32` and received field is `str` on which `parse_path` does not fail.

Let's not make that a part of this PR, but if you confirm that otherwise we don't need the kitchen sink of `parse_*` functions, I'll make the change in a followup PR.

Reply:
I thought the purpose of those functions is also validation of the data, but I suppose `dict_to_proto` handles that (and as you mentioned, it's maybe not even trezorlib's job)? Some validations would perhaps get lost, but nothing which isn't also handled in FW. It seems to me there isn't any special data handling/processing, which wouldn't allow the use of `dict_to_proto`.

@refi93 what do you think about this please?

Reply:
We discussed this offline with @refi93 and we don't see any issues with using `dict_to_proto` - it would simplify the parsing quite some bit. So if you could make that change, that would be awesome.

Also we realized that the same either/or structure is present in transaction outputs and it isn't correctly validated in the code - there's also just an if/elif there. So I'll add the `data_provided` pattern there also as a new commit here if that's OK. 

And I would rename the variable to `fields_provided` if that's also OK, it seems like a better fit to me.

Reply:
`fields_provided` added to auxiliary data validation https://github.com/trezor/trezor-firmware/pull/1564/commits/a2e97e9525c3f9afe7edf9ee89056ba766ebb69d

Reply:
On second thought, I won't add the output validation in this PR. We have another PR opened (just internally atm) which refactors address validation so I will add it there as it directly touches that code.